test_provider.py

```diff
@@ -0,0 +1,454 @@
+        TOTAL_TIME = 'totalTime'
+        OFFSET_P = 'telescope:p'
+        OFFSET_Q = 'telescope:q'
+    
+    class _TimingWindowsKeys(Enum):
+        TIMING_WINDOWS = 'timingWindows'
+        START = 'start'
+        DURATION = 'duration'
+        REPEAT = 'repeat'
+    
+    class _MagnitudeKeys(Enum):
+        NAME = 'name'
+        VALUE = 'value'
+
+    def __init__(self, path):
+        self.path = path
+
+    def load_program(self, path: str) -> dict:
+        with open(path, 'r') as f:
+            return json.loads(f.read())
+    
+    @staticmethod
+    def sort_observations(obs: List[Observation]) -> List[Observation]:
+        return sorted(obs, key=lambda x: x.order)
+    
+    @staticmethod
+    def strip_constraints(constraint: str):
+        return constraint.split('/')[0].split('%')[0]
+
+    @staticmethod
+    def parse_magnitude(json: dict) -> Magnitude:
+        band = MagnitudeBands[json['name']]
+        value = json['value']
+        error = None
+        return Magnitude(band, value, error)
+    
+    @staticmethod
+    def get_program_dates(program_type: ProgramTypes, id: str, notes: list) -> Tuple[datetime, datetime]:
+        start = end = None
+        year = int(id[3:7])
+        next_year = year + 1
+        semester = id[7]
+        if program_type is ProgramTypes.FT:
+
+            for note in notes:
+                if 'title' in note.keys():
+                    title = note['title'].lower()
+                    if 'cycle' in title or 'active' in title:
+                        fields = title.strip().split(' ')
+                        months = []
+                        for field in fields:
+                            if '-' in field:
+                                months = field.split('-')
+                                if len(months) == 3:
+                                    break
+                        if len(months) == 0:
+                            for field in fields:
+                                f = field.lower().strip(' ,')
+                                if f in ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec']:
+                                    months.append(f)
+                        im1 = months[0]
+                        im2 = months[-1]
+                        if semester == 'B' and im1 < 6:
+                            start = datetime(next_year, im1, 1)
+                            end = datetime(next_year, im2, 1)
+                        else:
+                            start = datetime(year, im1, 1)
+                            end = datetime(year, im2, 1) if im2 > im1 else datetime(next_year, im2, 1)
+                        break
+        else:
+            if semester == 'A':
+                start = datetime(int(year), 2, 1)
+                end = datetime(int(year), 7, 31)
+            else:
+                start = datetime(int(year), 8, 1)
+                end = datetime(int(next_year), 1, 31)
```

Comment:
If there is some kind of error in the `semester` as specified, it will never be caught, which is why I think it's important to work with `SemesterHalf` and do the lookup instead of comparing `str`. If we attempt to do a lookup on a `SemesterHalf` that doesn't exist, then a `KeyError` will automatically be `raise`d. Right now, if an illegal `semester` is encountered, it will just always be treated as `B`.

Reply:
I ended up working on this method a bit, but I think the logic was well simplified over the XML: it just wasn't clear to me when I started, and I did add some error handling and divided it up a bit to make it more easily readable, but good job on cleaning this up.